oak/server/rust/oak_runtime/src/node/grpc_server.rs

```diff
@@ -0,0 +1,195 @@
+    writer: Handle,
+    thread_handle: Option<JoinHandle<()>>,
+}
+
+impl GrpcServerNode {
+    /// Creates a new [`GrpcServerNode`] instance, but does not start it.
+    pub fn new(
+        config_name: &str,
+        runtime: RuntimeRef,
+        node_id: NodeId,
+        reader: Handle,
+        writer: Handle,
+    ) -> Self {
+        Self {
+            config_name: config_name.to_string(),
+            runtime,
+            node_id,
+            reader,
+            writer,
+            thread_handle: None,
+        }
+    }
+}
+
+impl super::Node for GrpcServerNode {
+    fn start(&mut self) -> Result<(), OakStatus> {
+        let service = GrpcService {
+            config_name: self.config_name.clone(),
+            runtime: self.runtime.clone(),
+            node_id: self.node_id,
+            reader: self.reader,
+            writer: self.writer,
+        };
+        // TODO(#770): Use `std::thread::Builder` and give a name to this thread.
+        let thread_handle = spawn(move || {
+            let pretty_name = format!("{}-{:?}:", service.config_name, thread::current());
+            // let address = format!("[::]:{}", port).parse()?;
+            let address = "127.0.0.1:8080".parse().unwrap();
+
+            let service_function = hyper::service::make_service_fn(move |_| {
+                let connection_service = service.clone();
+                async move {
+                    Ok::<_, hyper::Error>(hyper::service::service_fn(move |req| {
+                        let request_service = connection_service.clone();
+                        async move { request_service.serve(req).await }
+                    }))
+                }
+            });
+
+            let server = hyper::Server::bind(&address).serve(service_function);
+            if let Err(error) = block_on(server) {
+                error!("{} LOG: RPC server pseudo-node error: {}", pretty_name, error);
+            }
+
+            info!("{} LOG: exiting gRPC server thread", pretty_name);
+        });
+        self.thread_handle = Some(thread_handle);
+        Ok(())
+    }
+    fn stop(&mut self) {
+        if let Some(join_handle) = self.thread_handle.take() {
+            if let Err(err) = join_handle.join() {
+                error!("error while stopping gRPC server node: {:?}", err);
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+struct GrpcService {
+    config_name: String,
+    runtime: RuntimeRef,
+    node_id: NodeId,
+    reader: Handle,
+    writer: Handle,
+}
+
+impl GrpcService {
+    async fn serve(
+        &self,
+        req: hyper::Request<hyper::Body>,
+    ) -> Result<hyper::Response<hyper::Body>, hyper::Error> {
+        let body = hyper::body::aggregate(req)
+            .await
+            .expect("Couldn't get request body");
+
+        let client_message = crate::message::Message {
+            // TODO: Is it copying the data?
+            data: body.bytes().to_vec(),
+            channels: vec![],
+        };
```

Comment:
I'm passing received data to the Runtime as it is (just copy bytes from the HTTP body).
Is this correct?
cc @tiziano88 @daviddrysdale

Reply:
I don't think this is what the C++ runtime does (and what the SDK expects). Each incoming request should create a gRPC invocation message, which is a pair of channels (and no bytes), over which the actual request(s) can be read, and response(s) can be written. See https://github.com/project-oak/oak/blob/master/oak/server/invocation.h . Also, in your PR, how would clients reply to a particular request?

Reply:
Ok, will implement an invocation

Reply:
IIUC `grpc_request` that is implemented in `oak_tests` just passes bytes to the `writer`:
https://github.com/project-oak/oak/blob/589b2d20bbd9e0d8388f58e8b50b50522842013c/sdk/rust/oak_tests/src/lib.rs#L105-L124

Reply:
I think this is what gets sent to the node under test

https://github.com/project-oak/oak/blob/589b2d20bbd9e0d8388f58e8b50b50522842013c/sdk/rust/oak_tests/src/lib.rs#L133-L142